onnxruntime/contrib_ops/webgpu/quantization/subgroup_matrix_matmul_nbits.cc

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "contrib_ops/webgpu/quantization/subgroup_matrix_matmul_nbits.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace webgpu {
+
+Status SubgroupMatrixMatMulNBitsProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  shader.AddInput("input_a", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias | ShaderUsage::UseValueTypeAlias);
+  shader.AddInput("input_b", ShaderUsage::UseUniform);
+  shader.AddInput("scales_b", ShaderUsage::UseUniform);
+  shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseElementTypeAlias);
+
+  shader.AdditionalImplementation() << R"ADDNL_FN(
+        const tile_cols = 64;
+        const tile_rows = 32;
+        const tile_k = 32;
+        const subtile_cols = 32;
+        const subtile_rows = 16;
+        const quantization_block_size = 32;
+        alias compute_precision = output_element_t;
+
+        var<workgroup> tile_A: array<compute_precision, tile_rows * tile_k>;       // 32 x 32 - RxC
+        var<workgroup> tile_B: array<compute_precision, tile_cols * tile_k>;       // 64 x 32 - RxC
+        var<workgroup> scratch: array<array<array<compute_precision, 64>, 4>, 4>;  // 64 * 4 * 4
+
+        fn loadSHMA(tile_base: u32, k_idx: u32, row: u32, c_idx:u32) {
+            let a_global = tile_base + row;
+            if (a_global >= uniforms.M) {
+                return;
+            }
+            // Each call loads 8 columns, starting at col.
+            var col = c_idx * 8;
+            // 128 threads need to load 32 x 32. 4 threads per row or 8 col per thread.
```

Comment:
Will it be better if 32 threads per row so that the contiguous threads can access contiguous data in memory (each call loads 1 column, but needs to load 8 rows)?  like, `loadSHMA(a_global_base, kidx, local_idx/tile_k, local_idx%tile_k)` and change below loop for row offset instead of col offset. Usually, such kind of access pattern can bring better performance.

Reply:
100%, Ill work on this. Perhaps that can be a follow up change.

Reply:
Prototyped it here https://github.com/microsoft/onnxruntime/pull/23758

Just with the changes to LoadSHMA I see perf regression

(onnxruntime-genai) sushraja@sushrajas-Mac-mini c % ./model_benchmark -i ~/Phi3_Edge -l 1000                                      
Batch size: 1, prompt tokens: 1001, tokens to generate: 128
Prompt processing (time to first token):
        avg (us):       6.07132e+06
        avg (tokens/s): 164.874
        p50 (us):       6.07084e+06
        stddev (us):    4698.02
        n:              5 * 1001 token(s)

Reply:
Weird that it has regression. But thanks for the information.

Reply:
Hi, I see both `LoadSHMA` and `LoadSHMB` are changed in the prototype. Do you see perf gain with both changes applied?